### PR TITLE
Improving Client Unit Tests

### PR DESCRIPTION
**Pull request by ArquintL**
Includes changeset of PR #22

Takes several ideas from Prusti Assistant (thanks to @fpoli) and applies them to Gobra-IDE.
Most importantly, it executes the test suite in three different configurations (stable, nightly, and with an empty config file).
In addition, Gobra Tools get installed to a folder provided by VSCode. This ensures that uninstalling the extension will also delete the Gobra Tools. Furthermore, a confirmation popup appears to inform users that we download and install the Gobra Tools. Users have to confirm it.

### CODE REVIEW

File: client/src/test/index.ts [hunk 1]
```diff
@@ -21,39 +21,58 @@
 import * as path from 'path';
 import * as Mocha from 'mocha';
 import * as glob from 'glob';
+import NYC = require("nyc");
```
Comment (jcp19) [inline client/src/test/index.ts:24]: What an horrible name for a library :D

**Review by jcp19 — APPROVED** [review]
I identified some minor things that can be changed. Otherwise, this looks fine (although I have no experience in TS). I did not look into the `.gobra` files

File: client/src/Helper.ts [hunk 2]
```diff
@@ -217,32 +236,48 @@ export class Helper {
...
     return process.env["GITHUB_TOKEN"];
   }
 
+  /**
+   * Returns true if `getGobraToolsPath` should be wiped after activating the extension to ensure a clean system state.
+   */
+  public static cleanInstall(): boolean {
+    const value = process.env["GOBRA_IDE_CLEAN_INSTALL"];
+    return value != null && 
+      (value == "1" || value == "true" || value == "TRUE");
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply client/src/Helper.ts:249]: Great suggestion!

File: client/.vscodeignore [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
 evaluationResults.txt
-tmp
-**/*.js.map
-.vscode
-.vscode-test
-node_modules
-src/
+package-notes.md
+report.csv
+tsconfig.json
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply client/.vscodeignore:7]: Good point. Actually, `package-lock.json` seems to automatically be ignored but I made it now explicit anyway. `package-lock.json` should be committed as it specifies the versions of all transitive dependencies

File: client/src/test/index.ts (hunk 1 above)
Comment (ArquintL) [reply client/src/test/index.ts:24]: haha

File: client/src/evaluate/index.ts [hunk 4]
```diff
@@ -22,37 +22,35 @@ import * as path from 'path';
 import * as Mocha from 'mocha';
 import * as glob from 'glob';
 
-export function run(): Promise<void> {
+export async function run(): Promise<void> {
   // Create the mocha test
-  const mocha = new Mocha({
-    ui: 'tdd',
+	const mocha = new Mocha({
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply client/src/evaluate/index.ts:27]: It was a mixing of spaces and tabs ^^

File: client/src/Helper.ts [hunk 5]
```diff
@@ -19,19 +20,34 @@ export class Helper {
   public static isMac = /^darwin/.test(process.platform);
 
   public static isServerMode(): boolean {
-    return vscode.workspace.getConfiguration("gobraSettings").get("serverMode");
+    const mode = vscode.workspace.getConfiguration("gobraSettings").get<boolean>("serverMode");
+    if (mode) {
+      return mode;
+    } else {
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (ArquintL) [reply client/src/Helper.ts:28]: I wanted to check whether `mode` is null and return in this case `true` (as a default) but I messed it up ^^

**Comment by ArquintL** [comment]
The many `.gobra` files come from the evaluation of Silas' BSc thesis. Imho it makes more sense to store them in this repo than to forget about them on gitlab.

**Pull request merged**